### DISCUSSION TITLE
폴더 구조에 대한 토의

### DISCUSSION THREAD

**Discussion by Gyaak** [Kernel360/F2-BAGUNI#106] [개발 토의]
폴더는 폴더와 픽을 가지고 있을 수 있다.
폴더의 구조와 내부 순서는 json 형태로 저장된다.
폴더안에 들어있는 폴더/픽의 정보는 json을 표현되며 type 필드에 의해 구분된다
### 폴더구조json 의 각 원소들의 필드
- type : PICK | FOLDER
- data : 세부정보는 data 필드안에 존재하며, type에 따라 data의 구조는 다를 수 있다.
  1. type == PICK
      - data = { pickId, title, memo, tag, .. 등 카드에 표시되는 데이터 }
      - 사실 server에서는 pickId만 있으면 되기에, client에서 보내기 편한 방식으로 보내면 됨
  2. type == FOLDER
      - data = { folderId, name, children(json 배열) }
  - 여기서 pickId, folderId는 server측 id이다.

최초 화면을 그릴 때, server로부터 폴더구조json과 데이터를 받아 화면을 그린다.
그 이후에는 오류가 발생하지 않는 한 전체 폴더구조json을 응답으로 받지 않는다.
- 오류 발생시 오류 표시 후 server측 폴더구조json으로 화면을 새로 그린다.

미분류 폴더는 folderId = 0 으로 사용자마자 딱 1개 가질 수 있음.
-> **_미분류 폴더(최상위폴더)에 대해서는 추가 논의 필요_**
이 폴더는 회원가입시 기본적으로 생기는 폴더

---

픽 추가시 요청에 담겨야 하는 정보
- 픽에 대한 정보 (title, url, tag ....)
- 해당 픽이 속한 folderId (폴더 미지정시(미분류) folderId = 0)
- 해당 픽이 추가된 구조가 반영된 폴더구조json

이미 존재하는 픽의 폴더 변경시 요청에 담겨야 하는 정보
- pickId
- 해당 픽이 속한 folderId (폴더 미지정시(미분류) folderId = 0)
- 해당 픽이 변경된 구조가 반영된 폴더구조json

픽들의 순서가 변경시 요청에 담겨야 하는 정보
- 변경이 반영된 폴더구조json

픽 삭제시 요청에 담겨야 하는 정보
- pickId
- 해당 픽이 삭제된 구조가 반영된 폴더구조json
- 만약 삭제 실패시(400번대응답) 이전 json구조로 롤백..?
   -> @enigsuss  이게 가능한지 아니면 server에서 변경전 폴더구조json을 내려주는게 맞을 지 의견주세요
   -> 지금은 server에서 이전 구조를 받는 것이 더 좋을 것 같습니다.

픽 세부정보 변경시 요청에 담겨야 하는 정보
- pickId
- 픽 세부 데이터(title, url, tag, ...)

---

폴더 추가시 요청에 담겨야 하는 정보
- name
- 상위 폴더의 id(parentId) 최상위 폴더라면 parentId = 0
- 해당 폴더가 추가된 폴더구조json

이미 존재하는 폴더의 위치 변경시 요청에 담겨야 하는 정보
- folderId
- 상위 폴더의 id(parentId) 최상위 폴더라면 parendId = 0
- 해당 폴더의 변경이 반영된 폴더구조json

폴더의 순서가 변경된 경우 요청에 담겨야 하는 정보
- 변경이 반영된 폴더구조json

폴더 삭제시 요청에 담겨야 하는 정보
- folderId
- 해당 폴더의 삭제가 반영된 폴더구조json
  - 폴더 삭제시 안에 있는것들은 구조를 유지한채 미분류로 이동
- 만약 삭제 실패시(400번대응답) 이전 폴더구조json로 롤백..? (픽과 마찬가지)

폴더 이름 변경 시 요청에 담겨야 하는 정보
- folderId
- name

---

### 조회 메소드

- pickId 로 조회
- folderId 로 조회
  - 해당 폴더에 들어있는 폴더(id, name)와 픽(세부정보)을 json 배열 형태로 반환
  - 정렬x -> 프론트에서 가지고 있는 폴더구조json으로 정렬하여 사용

---

## 폴더 구조 예시 (JSON) 
@enigsuss 이렇게 바뀌었습니다.
```json
{
  "root": [
    { "id":"1", "type":"folder", "folderId":1, "name":"폴더", "children": [] },
    { "id":"2", "type":"folder", "folderId":2, "name":"폴더", "children": [] },
    { "id":"3", "type":"folder", "folderId":2, "name":"폴더", "children": [
      { "id":"4", "type":"folder", "folderId":2, "name":"폴더", "children": [] },
      { "id":"5", "type":"folder", "folderId":2, "name":"폴더", "children": [] }
    ] }
  ],
  "recycleBin": [],
  "unclassified": []
}
```

**Comment by sangwonsheep** [comment]
확인 완료했습니다. 고생하셨습니다!!

**Comment by kimminkyeu** [comment]
네 확인했습니다. 프론트측에서 보내주는 json 구조 는 `{ id(프론트 전용) , data }`이 되겠네요.
재용님이 폴더구조json 예시 적어주시면 바로 확인하겠습니다. 늦은 시간까지 고생하셨습니다 :+1:

**Comment by enigsuss** [comment]
JSON 데이터 예시 추가했습니다. 고생하셨습니다!

**Reply by kimminkyeu** [reply]
@enigsuss json 키를 `recycle_bin` 에서 `recycleBin`으로 변경 부탁드립니다. 안에는 다 카멜 케이스인데 쟤만 스네이크네요.
위 json 예시도 업데이트했습니다.

**Reply by enigsuss** [reply]
넵 추후 작업시 반영하겠습니다.
